### PR TITLE
Raise syntax error when method argument name is keyword

### LINKED ISSUE #5895: Don't allow begin, end as method arguments
I expect a compile-time error while using reserved `begin`, `end` keywords for method arguments unless they're used as external names. 

```crystal
def plan(begin, end)
  puts begin
  puts end
end

plan "Crystal", "External"
```

This just tries to execute the method :neutral_face:

### PR DESCRIPTION
**Pull request by MakeNowJust**
Closes #5895

However some keywords are valid argument name still.
Invalid argument names are listed here:

    begin nil true false yield with abstract
    def macro require case select if unless include
    extend class struct module enum while until return
    next break lib fun alias pointerof sizeof
    instance_sizeof typeof private protected asm
    end

Above names except `end` are handled as keyword by `Crystal::Parser.parse_atomic_without_location`.
It means, we cannot assign value into them and never reference them.
In other words we have no way using them, so it does not make sense to allow these names as argument name.

Additionally `end` is confusable. It maybe terminate `def` block.

Thank you.

### CODE REVIEW

File: spec/compiler/parser/parser_spec.cr [hunk 1]
```diff
@@ -163,6 +163,20 @@ describe "Parser" do
   assert_syntax_error "def foo!=; end", "unexpected token: !="
   assert_syntax_error "def foo?=(x); end", "unexpected token: ?"
 
+  # #5895
+  %w(
+    begin nil true false yield with abstract
```
Comment (oprypin) [inline spec/compiler/parser/parser_spec.cr]: The manual list is bad enough, let's not put in two places. Just add a few examples.

Comment (oprypin) [reply spec/compiler/parser/parser_spec.cr]: And don't use any interpolation in the examples because tests should be "unrolled"

File: spec/compiler/parser/parser_spec.cr [hunk 2]
```diff
@@ -163,6 +163,20 @@ describe "Parser" do
...
+    def macro require case select if unless include
+    extend class struct module enum while until return
+    next break lib fun alias pointerof sizeof
+    instance_sizeof typeof private protected asm
+    end
+  ).each do |kw|
+    assert_syntax_error "def foo(#{kw}); end", "cannot use '#{kw}' as argument name", 1, 9
+    assert_syntax_error "def foo(foo #{kw}); end", "cannot use '#{kw}' as argument name", 1, 13
+    it_parses "def foo(#{kw} foo); end", Def.new("foo", [Arg.new("foo", external_name: kw.to_s)])
```
Comment (oprypin) [inline spec/compiler/parser/parser_spec.cr]: Nice to see that external names are not forgotten

File: spec/compiler/parser/parser_spec.cr (hunk 1 above)
Comment (MakeNowJust) [reply spec/compiler/parser/parser_spec.cr]: Why bad enough? `parser_spec.cr` uses such a list in 10 places at least. And I found the implementation bug by this test before commiting in fact.

File: src/compiler/crystal/syntax/parser.cr [hunk 3]
```diff
@@ -3588,6 +3592,10 @@ module Crystal
 
       case @token.type
       when :IDENT
+        if @token.keyword? && invalid_internal_name?(@token.value)
+          raise "cannot use '#{@token}' as argument name", @token
```
Comment (Sija) [inline src/compiler/crystal/syntax/parser.cr]: `... as an argument name`?

Comment (MakeNowJust) [reply src/compiler/crystal/syntax/parser.cr]: Good :+1: :100: :tada:

**Comment by Sija** [comment]
How about prohibiting use of such keywords as variable names too?

File: src/compiler/crystal/syntax/parser.cr [hunk 4]
```diff
@@ -3648,6 +3659,24 @@ module Crystal
       {arg_name, external_name, found_space, uses_arg}
     end
 
+    def invalid_internal_name?(keyword)
+      case keyword
+      # These names are handled as keyword by `parser_atomic_without_location`.
```
Comment (bcardiff) [inline src/compiler/crystal/syntax/parser.cr]: The reference should be `parse_atomic_without_location` (without the `r`). And I would add a note in `Parser#parse_atomic_without_location` as well to keep both list in sync.

Comment (MakeNowJust) [reply src/compiler/crystal/syntax/parser.cr]: Thank you. I'll fix it.

**Comment by MakeNowJust** [comment]
@Sija At first we cannot use such keywords as variable names in reality currently.

However there is an exception: `foo : Type` syntax. For example, following code is valid for now.

```crystal
begin : Int32
```

This syntax is used with a macro normally. For example:

```crystal
property begin : Int32
```

The first case should be prohibited and the second should be allowed. But I don't know how to implement this. Sorry.
